### PR TITLE
Hashcat

### PR DESCRIPTION
**Pull request by andreea-popescu-reef**
added synthetic job that sequentially does SHA algorithms on multiple hashes

### CODE REVIEW

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/gpu_hashcat.py [hunk 1]
```diff
@@ -1,44 +1,42 @@
...
-        elif algo == Algorithm.SHA512:
-            return 5, "1710", 45
-        else:
-            raise ValueError(f"Invalid algorithm: {algo}")
+        self.algorithm = Algorithm.get_random_algorithm()
+        self.hashcat_hash_type = self.algorithm.type
+
+        # set synthetic_jobs difficulty depending on subnet weigfhts_version
+        num_hashes, num_digits, num_letters = None, None, None
+        if get_subnet_weights_version() > 0:
```
Comment (mzukowski-reef) [inline validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/gpu_hashcat.py:31]: I feel that num_hashes should be part of a job params description.
We should have multiple sets of possible param values that are assigned to minimum weights_version, we have to make it backwards compatible so set for weights_version == 0 current parameter values - and for "next version" (is it `1`?)  should have new values (num_hashes=50 etc) and we choose max version from those that are "lte" current weights_version (so we can increase this param without need of code change)

also we have to benchmark new params (with num_hashes) and set proper timeout - scoring function now is linear from 2 to -inf with 2 in t=0, 1 in t=timeout and 0 in t=2*timeout

we should at least now how we changing scoring - I mean if same hardware would get same score after change or we introduce some "difficulty pressure"

**Review by agoncharov-reef — APPROVED** [review]
I read it and haven't spotted anything critical, although I don't have in-depth knowledge in the topic

**Review by agoncharov-reef — APPROVED** [review]
I read it and haven't spotted anything critical, although I don't have in-depth knowledge in the topic

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/synthetic_job.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        return self.params[self]["hash_type"]
+
+    @staticmethod
+    def get_random_algorithm():
+        algorithms = list(Algorithm)
+        return algorithms[datetime.datetime.utcnow().hour % len(algorithms)]
+
+    @staticmethod
+    def get_all_algorithms():
+        return list(Algorithm)
```
[the remark this answers is not in the document]
Comment (ppolewicz) [reply validator/app/src/compute_horde_validator/validator/synthetic_jobs/synthetic_job.py:60]: @andreea-popescu-reef https://github.com/reef-technologies/python-review-zoo/?tab=readme-ov-file#usage-of-staticmethod

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/v1_decrypt.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                payload = Fernet(key).decrypt(payload).decode("utf-8")
+
+            with open("_payload.txt", mode="wb") as f:
+                f.write(payload.encode("utf-8"))
+
+            cmd = f'hashcat --potfile-disable --restore-disable --attack-mode 3 --workload-profile 3 --optimized-kernel-enable --hash-type {algorithm} --hex-salt -1 "?l?d?u" --outfile-format 2 --quiet _payload.txt "{mask}"'
+            passwords = subprocess.check_output(cmd, shell=True, text=True)
+            passwords = sorted(passwords.split('\n'))[1:]
+            answers.append(passwords)
+        print(answers)
```
[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply validator/app/src/compute_horde_validator/validator/synthetic_jobs/v1_decrypt.py:65]: no point :|
yes that should be a variable

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/v1_synthetic_job.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        cls, algorithms: list[Algorithm], params: list[JobParams], salt_length_bytes: int = 8
+    ) -> Self:
+
+        # generate distinct passwords for each algorithm
+        passwords = []
+        for _params in params:
+            _passwords = set()
+            while len(_passwords) < _params.num_hashes:
+                _passwords.add(cls.random_string(_params.num_letters, _params.num_digits))
+            passwords.append(sorted(list(_passwords)))
```
[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply validator/app/src/compute_horde_validator/validator/synthetic_jobs/v1_synthetic_job.py:47]: if I do it like that I'll get less than `num_hashes` passwords when the random_string generates a duplicate

**Review by mzukowski-reef — APPROVED** [review]

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/v1_decrypt.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        with open("_payload.txt", mode="wb") as f:
+            f.write(payload.encode("utf-8"))
+
+        cmd = f'hashcat --runtime {timeout} --potfile-disable --restore-disable --attack-mode 3 --workload-profile 3 --optimized-kernel-enable --hash-type {algorithm} --hex-salt -1 "?l?d?u" --outfile-format 2 --quiet _payload.txt "{mask}"'
+        passwords = subprocess.check_output(cmd, shell=True, text=True)
+        passwords = [p for p in sorted(passwords.split('\n')) if p != '']
+        answers.append(passwords)
+
+    print(
+        b64encode(hashlib.sha256(str(answers).encode("utf-8")).digest(), altchars=b"-_").decode("utf-8")
```
[the remark this answers is not in the document]
Comment (andreea-popescu-reef) [reply validator/app/src/compute_horde_validator/validator/synthetic_jobs/v1_decrypt.py]: changed

**Pull request merged**